### ISSUE TITLE
Return list of PDB IDs for a given gene

### ISSUE THREAD

**Issue by eweitz** [NCBI-Hackathons/PDB2Genome#3]
A use case mentioned in [our README](https://github.com/NCBI-Hackathons/PDB2Genome/tree/2436ff6ac9e7e3317d553553a4f806e6aca4b251#pdb2genome) is:

> **Mapping PDB structures to gene-level** in genome browsers, and visualizing SNPs

We currently have lists of PDB files (somewhat grouped by chromosomes), but it seems we should aim to be able to show PDB structures grouped by gene, e.g. for gene [C1QA](http://www.ncbi.nlm.nih.gov/gene/712) (NCBI Gene ID 712).  Perhaps something like:

```
$ cat pdbs_for_gene/C1QA.txt
2WNU_A
2WNU_D
2WNV_A
2WNV_D
2HKJ_A
5HZF_A
```

I don't know how to get a list of PDB IDs (PDB_id, PDB_chain tuples?) for a given gene symbol off the top of my head.  But I suspect we could use EUtils to:
1. Get Gene ID for gene symbol (C1QA -> 712).
2. UniProt ID(s) for Gene ID (712 -> P02745)
3. PDB IDs for UniProt ID (P02745 -> 2WNU_A, 2WNU_D, 2WNV_A, 2WNV_D, 2HKJ_A, 5HZF_A)

I can easily do (1) and (2), but at the moment I don't know how to do (3).  Perhaps (3) could be done by someone else.

**Comment by DCGenomics** [comment]
I'm fairly certain #3 is in the parquet files we have been parsing!

**Comment by pkerpedjiev** [comment]
What format would you like it in?

```
cat pdbs_for_uniprot/P17542.txt  
2YPA_A
2YPB_A
```

**Comment by eweitz** [comment]
That looks great.  

Although I could do parts (1) and (2) outlined in the original description, it occurs to me that I'm probably the person best suited on our team to develop Issue #2, which is a higher priority, and will likely take me much of the rest of today.

If anyone else on the team has spare cycles and is interested in exploring the NCBI EUtils API, then having another person take on parts (1) and (2) above might be worthwhile.

**Comment by pkerpedjiev** [comment]
(1) Isn't the gene symbol - gene id mapping in some NCBI table somewhere?
(3) is in `/home/ubuntu/pkerp/output/uniprot_pdbs`

**Comment by eweitz** [comment]
> (1) Isn't the gene symbol - gene id mapping in some NCBI table somewhere?

Yes, such mappings exist in many public NCBI resources.  Examples include the NCBI's EUtils web API and FTP files available from NCBI [Gene](http://ftp.ncbi.nih.gov/gene/DATA/GENE_INFO/Mammalia/) or NCBI [Eukaryotic Genome Annotation Pipeline](http://ftp.ncbi.nih.gov/genomes/Homo_sapiens/GFF/).  

I gravitate toward web APIs.  Below is how to use NCBI EUtils to get UniProt IDs for a gene symbol.  

# Get Gene ID for gene symbol

Gene symbol: [C1QA](http://www.ncbi.nlm.nih.gov/gene/712)

Request:
http://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?db=gene&retmode=json&term=C1QA[PREF]%20AND%20Homo%20sapiens[ORGN]

Response:

```
{
    "header": {
        "type": "esearch",
        "version": "0.3"
    },
    "esearchresult": {
        "count": "1",
        "retmax": "1",
        "retstart": "0",
        "idlist": [
            "712"
        ],
```

The Gene ID is 712, and can be parsed via `response["esearchresult"]["idlist"][0]`.

# Get UniProt ID for Gene ID

Gene ID: 712

Request (sorry, I don't know why this isn't available in JSON):
http://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=gene&retmode=xml&retformat=text&id=712

```
$ curl "http://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=gene&retmode=xml&retformat=text&id=712" > gene_report_712.xml
$ grep -A 3 "<Dbtag_db>UniProtKB/Swiss-Prot</Dbtag_db>" gene_report_712.xml
                                  <Dbtag_db>UniProtKB/Swiss-Prot</Dbtag_db>
                                  <Dbtag_tag>
                                    <Object-id>
                                      <Object-id_str>P02745</Object-id_str>
--
                      <Dbtag_db>UniProtKB/Swiss-Prot</Dbtag_db>
                      <Dbtag_tag>
                        <Object-id>
                          <Object-id_str>P02745.2</Object-id_str>
```

The UniProt ID there is, of course, P02745.  That could be parsed out using XPath, via e.g. lxml in Python.

With that approach, the gene symbol C1QA could be mapped to the UniProt ID P02745.  Fortunately we already have the link between UniProt ID and PDB IDs.

For what it's worth, my rationale for thinking we need to this is that use case: "**Mapping PDB structures to gene-level** in genome browsers".  Genes tend to be the focal point of genome browsers, so looking up PDB structures given gene symbol seems like it would be the easiest link to those PDB BED files we have.

**Comment by eweitz** [comment]
> (3) is in /home/ubuntu/pkerp/output/uniprot_pdbs

Looks good!

**Comment by windlust01** [comment]
Two search steps for a gene ID (ie. LPL = 4023)  to related PDB structures.
**1) from gene to RefSeq Protein**
https://eutils.ncbi.nlm.nih.gov/entrez/eutils/elink.fcgi?dbfrom=gene&db=protein&id=4023&linkName=gene_protein_refseq
`
<eLinkResult>
<LinkSet>
<DbFrom>gene</DbFrom>
<IdList>
<Id>4023</Id>
</IdList>
<LinkSetDb>
<DbTo>protein</DbTo>
<LinkName>gene_protein_refseq</LinkName>
<Link>
<Id>4557727</Id>
</Link>
</LinkSetDb>
</LinkSet>
</eLinkResult>`

**2) From RefSeq Protein to Related Structure.**
https://eutils.ncbi.nlm.nih.gov/entrez/eutils/elink.fcgi?dbfrom=protein&db=structure&id=4557727&linkName=protein_structure_related_list

`
<eLinkResult>
<LinkSet>
<DbFrom>protein</DbFrom>
<IdList>
<Id>4557727</Id>
</IdList>
<LinkSetDb>
<DbTo>structure</DbTo>
<LinkName>protein_structure_related_list</LinkName>
<Link>
<Id>50149</Id>
</Link>
<Link>
<Id>50148</Id>
</Link>
<Link>
<Id>21535</Id>
</Link>
<Link>
<Id>1567</Id>
</Link>
<Link>
<Id>1566</Id>
</Link>
<Link>
<Id>46640</Id>
</Link>
<Link>
<Id>1236</Id>
</Link>
<Link>
<Id>61349</Id>
</Link>
<Link>
<Id>56257</Id>
</Link>
<Link>
<Id>45290</Id>
</Link>
<Link>
<Id>48948</Id>
</Link>
<Link>
<Id>5534</Id>
</Link>
<Link>
<Id>51201</Id>
</Link>
<Link>
<Id>40178</Id>
</Link>
<Link>
<Id>11978</Id>
</Link>
</LinkSetDb>
</LinkSet>
</eLinkResult>`
